src/libraries/System.Net.Http/ref/System.Net.Http.cs

```diff
@@ -299,6 +301,11 @@ public sealed class HttpRequestOptions : System.Collections.Generic.IDictionary<
         public void Set<TValue>(HttpRequestOptionsKey<TValue> key, TValue value) { throw null; }
     }
 
+    public static class HttpRequestOptionsExtensions
+    {
+        public static void SetCustomMetricsTags(this HttpRequestOptions options, IReadOnlyCollection<KeyValuePair<string, object?>> tags) { throw null; }
```

Comment:
> The extension method should sit off `HttpRequestMessage`. From the dev's point of view, they're adding tags to the request.

I prefer not to pollute the `HttpRequestMessage` API with this niche feature. We came to a conclusion in #86281 with @noahfalk that having extensions on `HttpRequestOptions` would be a nice way to achieve this.

> I think the extension method is better for adding a single item.

I prefer the BCL extension methods to only deal with the bare minimum: setting/getting the collection in a safe way, so users have a strongly-typed API instead of copy-pasting the `HttpRequestOptionsKey` to their code. 

> There needs to be a way to add new key/values.

I would just change `IReadOnlyCollection` to `ICollection` and make the getter public. With that, middleware libraries could implement the rest in the way they prefer, eg. if someone doesn't need a mutable collection, they can pass an array which allocates less.

Reply:
I think if we are going to have an extensibility point using collections we should have either:
1. Actors can always set a collection type of their choice
2. All actors agree on a shared type (that is probably defined by the runtime).

As-is the design says there is only one collection but potentially multiple actors using it so (1) doesn't work. Unless we are going to restrict the interface to a single actor or add one collection per-actor I think the runtime needs to define the collection type that gets used.

Reply:
@noahfalk I have updated #86281 to reflect the requirement for adding tags from multiple handlers PTAL.

I don't really follow your concerns. Usage examples in #86281 demonstrate how to implement an extension that could be used from multiple handlers diagnostic events etc. (Is this what you mean by "actors"?) If you think your concerns apply for the updated proposal, can you please elaborate with an example?

Reply:
Np. By 'actor' I just meant any piece of software interacting with these APIs that has no specialized knowledge of the other pieces. For example developer A might write some code that creates HttpRequests, developer B might write a middleware handler, and developer C writes another handler. We can assume A, B, and C are unaware of each other's existence and put their libraries up on NuGet. Then an app developer came along and referenced all three things so they could use them together. In that case I'd say there are 3 actors (the code components written by A, B, and C respectively). 

> Usage examples in https://github.com/dotnet/runtime/issues/86281 demonstrate how to implement an extension that could be used from multiple handlers diagnostic events etc

Unless SetCustomMetricTags() is doing something more complex than I expected I don't think your example works as written? If EnricherHandler runs first to set the tags collection and then AddCustomTag() is called afterwards the call to tags.Add() fails with a NotSupportedException. The first actor decided to make a read-only collection and the second actor assumed the collection wouldn't be read-only. Either the API has to enforce that only certain types of collections are allowed, or the docs have to guide people to the right behavior and we hope they don't introduce those types of incompatibilities by mistake. 

Backing up a bit, is a lot of this concern trying to minimize the API surface having to do with avoiding extra allocations? A couple suggestions that you might want to look at with James:

1. An option that is close to your existing API and doesn't create dictionary entries or collections unless someone actually inserts a tag.

```C#
static TagList GetCustomMetricTags(this HttpRequestOptions options)
{
    if(options.TryGetValue(TagsKeyName, out TagList tags)
    {
        return tags;
    }
    return new TagList();
}

static void SetCustomMetricTags(this HttpRequestOptions options, TagList tags)
{
    if(tags.Count > 0 || options.ContainsKey(TagsKeyName))
    {
        options[TagsKeyName] = tags;
    }
}
```

2. Handling the extensibility differently than we've coming at it so far. From my POV the only requirement I care about is that R9 can add tags when they need to. Anything else about API being available specifically at the point an HttpRequest is being initialized or to middleware handlers I put extremely little weight on. If removing those constraints helped you make an API you were happier with or alleviated performance concerns I have no qualms with it. (I'm not speaking for @JamesNK who may have more specific rationale why certain requirements are important)
A MetricsOptions type hanging off the factory, a loose HttpMetricOptions type in the service collection, or a new property on the existing HttpClientFactoryOptions could all let R9 (or anyone else) register a callback that does the tag enrichment.

```C#
public delegate void ConfigureMetricTags(HttpRequest request, HttpResponse? response, ref TagList tags);

public class HttpMetricsOptions
{
    // Each callback in this list gets called once per request to modify the TagList
    // however they wish. I suspect this is about as good as it gets for CPU latency
    // and minimizing allocations per-request.
    public List<ConfigureMetricTags> ConfigureMetricTagCallbacks { get; }
}

internal class MetricHandler : DelegatingHandler
{
    private static void ApplyExtendedTags(ref TagList tags, HttpRequestMessage request, HttpResponseMessage? response)
    {
        // ...

        foreach(var callback in _options.RequestTagCallbacks)
        {
            callback(request, response, ref tags);
        }
    }
}
```

Reply:
> The first actor decided to make a read-only collection and the second actor assumed the collection wouldn't be read-only. 

I meant that extension methods in downstream libraries could deal with these kind of concerns. Since I received pushback on my approach both from you and James, I updated #86281 so we expose a single `GetCustomMetricsTags` method, PTAL.

Your second idea is interesting, but it seems to allocate more, and I would prefer to keep this API as simple as possible and move on with the proposal in the next API review.

Reply:
Just confirming for anyone else reading along that Anton and I chatted the other day and we agreed his current proposal is the simplest path forward for now. If we need it in the future the 2nd option I proposed could be added later and would be complimentary to any tags stored in the RequestOptions.